src/main/java/io/cdap/plugin/google/sheets/sink/GoogleSheetsRecordWriter.java

```diff
@@ -0,0 +1,373 @@
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Writes {@link FlatternedRowsRecord} records to Google drive via {@link GoogleDriveSinkClient}.
+ */
+public class GoogleSheetsRecordWriter extends RecordWriter<NullWritable, FlatternedRowsRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(GoogleSheetsRecordWriter.class);
+
+  private final Map<String, Map<String, Integer>> availableSheets = new HashMap<>();
+  private final Map<String, String> availableFiles = new HashMap<>();
+
+  private final Queue<FlatternedRowsRequest> recordsQueue = new ConcurrentLinkedQueue<>();
+  private final Map<String, Map<String, Integer>> sheetsContentShift = new HashMap<>();
+  private final Map<String, Map<String, Integer>> sheetsRowCount = new HashMap<>();
+  private final Map<String, Map<String, Integer>> sheetsColumnCount = new HashMap<>();
+  private final Semaphore threadsSemaphore;
+  private final ExecutorService writeService;
+  private final Semaphore queueSemaphore = new Semaphore(1);
+  private final int flushTimeout;
+  private GoogleSheetsSinkClient sheetsSinkClient;
+  private GoogleSheetsSinkConfig googleSheetsSinkConfig;
+  private ScheduledExecutorService formerScheduledExecutorService;
+  private boolean stopSignal;
+
+  public GoogleSheetsRecordWriter(TaskAttemptContext taskAttemptContext) throws IOException {
+    Configuration conf = taskAttemptContext.getConfiguration();
+    String configJson = conf.get(GoogleDriveOutputFormatProvider.PROPERTY_CONFIG_JSON);
+    googleSheetsSinkConfig =
+      GoogleSheetsOutputFormatProvider.GSON.fromJson(configJson, GoogleSheetsSinkConfig.class);
+
+    sheetsSinkClient = new GoogleSheetsSinkClient(googleSheetsSinkConfig);
+
+    writeService = Executors.newFixedThreadPool(googleSheetsSinkConfig.getThreadsNumber());
+    threadsSemaphore = new Semaphore(googleSheetsSinkConfig.getThreadsNumber());
+
+    flushTimeout = googleSheetsSinkConfig.getFlushExecutionTimeout();
+
+    formerScheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+    formerScheduledExecutorService.schedule(new TasksFormer(true),
+      flushTimeout, TimeUnit.SECONDS);
+
+  }
+
+  @Override
+  public void write(NullWritable nullWritable, FlatternedRowsRecord record) throws IOException, InterruptedException {
```

Comment:
if there are logically independent parts here, try to move to methods for readability.

Reply:
Added two separate methods and two separate classes